ksonnet-gen/ksonnet/catalog.go

```diff
@@ -0,0 +1,184 @@
+package ksonnet
+
+import (
+	"strings"
+
+	"github.com/go-openapi/spec"
+	"github.com/pkg/errors"
+)
+
+var (
+	blockedReferences = []string{
+		"io.k8s.apimachinery.pkg.apis.meta.v1.ListMeta",
+		"io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps",
+		"io.k8s.apimachinery.pkg.apis.meta.v1.Status",
+	}
+
+	blockedPropertyNames = []string{
+		"status",
+		"$ref",
+		"$schema",
+		"JSONSchemas",
+		"apiVersion",
+		"kind",
+	}
+)
+
+// ExtractFn is a function which extracts properties from a schema.
+type ExtractFn func(*Catalog, map[string]spec.Schema) (map[string]Field, error)
+
+// CatalogOpt is an option for configuring Catalog.
+type CatalogOpt func(*Catalog)
+
+// CatalogOptExtractProperties is a Catalog option for setting the property
+// extractor.
+func CatalogOptExtractProperties(fn ExtractFn) CatalogOpt {
+	return func(c *Catalog) {
+		c.extractFn = fn
+	}
+}
+
+// Catalog is a catalog definitions
+type Catalog struct {
+	apiSpec   *spec.Swagger
+	extractFn ExtractFn
+}
+
+// NewCatalog creates an instance of Catalog.
+func NewCatalog(apiSpec *spec.Swagger, opts ...CatalogOpt) (*Catalog, error) {
+	if apiSpec == nil {
+		return nil, errors.New("apiSpec is nil")
+	}
+
+	c := &Catalog{
+		apiSpec:   apiSpec,
+		extractFn: extractProperties,
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c, nil
+}
+
+// Resources returns a slice of all resources.
+func (c *Catalog) Resources() ([]Resource, error) {
+	var resources []Resource
+
+	for name, schema := range c.definitions() {
+		desc, err := ParseDescription(name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "parse description for %s", name)
+		}
+
+		component := NewComponent(schema)
+		if component == nil {
```

Comment:
It seems a bit odd to check whether the component is `nil`, when it is just got initialized from `NewComponent()`. Maybe `NewComponent()` should return an `err` as well. Thoughts?

Reply:
It did feel odd. I’m going to update it to run an error and then evaluate on the error type.